distribution/lib/Standard/AWS/0.0.0-dev/src/Internal/S3_File_Write_Strategy.enso

```diff
@@ -0,0 +1,66 @@
+private
+
+from Standard.Base import all
+import Standard.Base.Errors.Common.Forbidden_Operation
+import Standard.Base.Errors.File_Error.File_Error
+from Standard.Base.System.File.Generic.File_Write_Strategy import File_Write_Strategy, default_overwrite, default_append, default_raise_error, generic_remote_write_with_local_file
+
+import project.Errors.S3_Error
+
+## PRIVATE
+instance =
+    File_Write_Strategy.Value default_overwrite default_append default_raise_error s3_backup create_dry_run_file remote_write_with_local_file
+
+## PRIVATE
+create_dry_run_file file copy_original =
+    _ = [file, copy_original]
+    Error.throw (Forbidden_Operation.Error "Currently dry-run is not supported for S3_File, so writing to an S3_File is forbidden if the Output context is disabled.")
+
+## PRIVATE
+remote_write_with_local_file file existing_file_behavior action =
+    if existing_file_behavior == Existing_File_Behavior.Append then Error.throw (S3_Error.Error "S3 does not support appending to a file. Instead you may read it, modify and then write the new contents." Nothing) else
+        generic_remote_write_with_local_file file existing_file_behavior action
+
+## PRIVATE
+   A backup strategy tailored for S3.
+   Since S3 does not support a cheap 'move' operation, the standard backup
+   strategy that mostly relies on it does not work too well.
+   Instead, S3 relies on a simpler strategy:
+   1. If the destination file exists, copy it to a backup location
+      (overwriting a previous backup file, if it was present).
+   2. Write the new file.
+     i. If the write succeeded, that's it.
+        ii. If the write failed, 'restore' from the backup - copy the backup
+            back to the original location and delete the backup file (as it's no
+            longer needed because the original file is back with the old contents).
+s3_backup file action = recover_errors <|
```

Comment:
Could we integrate this into a single "File_Strategy" and allow us to choose what method is used based on compatibility

```
Type File_Strategy
    Value append_supported:Boolean=True move_supported:Boolean=True
```

I think for example OneDrive etc might work in similar ways to S3 but might have some more allowed APIs.

Reply:
That was my initial idea too. But I have abandoned it because I thought it is not flexible enough.

I cannot predict what limitations future backends (also ones added by external libs) will have. There may be backend-specific quirks in how we want to sometimes 'take a shortcut'.

I think the current approach with custom callbacks offers greatest flexibility and we can always re-use a common callback (like we do with the `default_overwrite` functions) - I guess maybe I should move the `s3_backup` strategy into `Base`.

On the other hand, looking at the _current_ state, I guess indeed the simpler approach with the boolean flags may work as well. I'm still not sure since it really depends a lot on what other backends bring - if all are rather simple the flag approach may work, I thought the current one may be more flexible (but it may also be too complex, it was created through quite a few iterations).

Reply:
Maybe let's keep it for this PR but I will add a TODO and we can revisit? What do you think?

Reply:
> This was originally posted on the wrong thread... sorry for that - reposting

I'm still undecided.

It feels like indeed the boolean flag could be enough for us at the moment. It seems to be enough for S3 and likely it will still be enough for some other simple backends.

I'm still not sure if it's better than the current approach which (at least in theory...) gives the (hypothetical) new backends more flexibility in how the details of each operation are carried out.

I have a hard time judging if I don't want to switch because I'm attached to my code and the sunken-cost fallacy OR if it's actually good intuition that this kind of flexibility can be useful in the future.

I have a feeling that the current more 'class-based' approach may be more readable (as we explicitly state how the implementation behaves, instead of setting some flags and 'deriving' an implementation based on the combination of these flags' values). But I know I may not be objective here.

-----

Given my indecisiveness here, I'd lean on keeping the code as-is for now - given that I think both approaches are essentially OKayish and that I prefer the current one and that keeping it is less work than rewriting - I think I'd just keep it for now.

But I'm happy to try a rewrite if requested.

Reply:
I think my general answer to this question is "you should design for what you have today" and not try to add abstractions for things you think you are going to need in the future. At the point you need the abstraction is the time to introduce it because then you actually have the concrete other thing that you need the abstraction for and can design the API to fit.

In this example if you had a concrete second backend you would know if the flag approach was good enough. So here I would invoke the YAGNI principle ("Always implement things when you actually need them, never when you just foresee that you [will] need them.") (https://en.wikipedia.org/wiki/You_aren%27t_gonna_need_it) and go for the simpler option.

Whether it is worth backing it now you have already written the more complex approach? I'm not sure...

Reply:
> I think my general answer to this question is "you should design for what you have today" and not try to add abstractions for things you think you are going to need in the future.

That's fair and I may have over-complicated this.

OTOH I'm not sure if the argument of being-open for future extension is the only one. I think it's also matter of whose responsibility it is to specify the way operations like `Backup` behave - should it be a generic implementation that selects the right strategy based on supported capabilities (the boolean flags) OR should that be the responsibility of each backend.

To me, it seemed simpler to give _the backend_ control over it. One of the arguments was future extensibility, but the other was that I can see in one place exactly how my backend will handle things, instead of having to look at how its capability flag settings influence what strategy is chosen.



> Whether it is worth backing it now you have already written the more complex approach? I'm not sure...

Yeah, I think I will keep it for now. Moreover, it seems like I may extend it a bit as part of #8833 and that may actually provide more rationale for keeping it like it is currently.

Reply:
I also think keep this for now. It's not really a choice between capability flags and backend implementation -- just like for the database tables, we have flags in some cases, and backend-specific implementations in other cases; it depends on the specific details of each backend and how much they have in common with each other.